### PR TITLE
Token Storage + UI

### PR DESCRIPTION
**Pull request by mohssenfathi**
## Description
This PR does two main things:
1. Adds a new TokenManager and supporting KeychainUtility so that we can save access tokens upon successful authentication.
2. Adds replacement UI elements for the UberButton + LoginButton.

### Supported platforms v13 -> v15
Bumped the minimum supported platform version to 15 to support the UIButton APIs. Platform adoption should be at a high enough level to support this. Also, this is in line with the 1P Uber app's min OS version.

### AccessToken
A new AccessToken struct has be created to accompany the new Auth stack. The old one has been marked as DEPRECATED, but the logic should be similar.

### Deprecated Classes
Legacy AccessToken and TokenManager have been marked as deprecated and call sites have been updated. There will be a later PR that migrates these call sites to the new classes.

### Saving AccessToken
AuthorizationCodeAuthProvider now contains logic to save the AccessToken to the keychain after successfully exchanging authorization code for access token.

### TokenManager + KeychainUtility
These are similar to the existing ones, with some improvements to the API. They now accept any objects conforming to the Codable protocol instead of NSCoding. Other logic should remain the same.

### UberButton + LoginButton
The existing UberButton and LoginButton classes have been replaced with updated APIs that are now available. The UI and logic remains the same. Some resources including colors and images have been added to the bundle to support the new buttons.

### Unit Tests
Unit tests have been added for new logic around token storage.

### Manual Testing
The UberButton/LoginButton has been added to the sample app for manual testing purposes.

### CODE REVIEW

File: Sources/UberCore/Authentication/Tokens/AccessToken.swift [hunk 1]
```diff
@@ -22,127 +22,45 @@
...
- via an OAuth representation.
- */
-public class AccessToken: NSObject, NSCoding {
-    /// String containing the bearer token.
-    public private(set) var tokenString: String
+public struct AccessToken: Codable, Equatable {
     
-    /// String containing the refresh token.
-    public private(set) var refreshToken: String?
+    public let accessToken: String?
```
Comment (alanpaivaa) [inline Sources/UberCore/Authentication/Tokens/AccessToken.swift:30]: Seems a bit odd having a `accessToken` property within a `AccessToken` type. Should we rename `AccessToken` maybe to `OAuthTokenInfo`?

File: Sources/UberCore/Authentication/Tokens/KeychainUtility.swift [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+/// @mockable
+public protocol KeychainUtilityProtocol {
+    
+    /// Saves an object in the on device keychain using the supplied `key`
+    ///
+    /// - Parameters:
+    ///   - value: The object to save. Must conform to the Codable protocol.
+    ///   - key: A string value used to identify the saved object
+    /// - Returns: A boolean indicating whether or not the save operation was successful
+    func save<V: Codable>(_ value: V, for key: String) -> Bool
```
Comment (alanpaivaa) [inline Sources/UberCore/Authentication/Tokens/KeychainUtility.swift:37]: We likely want `V` to just conform to `Encodable` here.

File: Sources/UberCore/Authentication/Tokens/KeychainUtility.swift [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+    
+    /// Retrieves an object from the on device keychain using the supplied `key`
+    ///
+    /// - Parameters:
+    ///   - key: The identifier string used when saving the object
+    /// - Returns: If found, an optional type conforming to the Codable protocol
+    func get<V: Codable>(key: String) -> V?
```
Comment (alanpaivaa) [inline Sources/UberCore/Authentication/Tokens/KeychainUtility.swift:44]: Ditto, but `Decodable`

File: Sources/UberCore/Authentication/Tokens/TokenManager.swift [hunk 4]
```diff
@@ -22,171 +22,93 @@
...
-        let success = keychainWrapper.deleteObjectForKey(identifier)
-        if success {
-            NotificationCenter.default.post(name: Notification.Name(rawValue: tokenManagerDidDeleteTokenNotification), object: self)
-        }
-        return success
+    /// Retrieves an Access Token from the on device keychain
+    ///
+    /// - Parameter identifier: The identifier string used when saving the Access Token
+    /// - Returns: An optional Access Token if found
+    @discardableResult
```
Comment (alanpaivaa) [inline Sources/UberCore/Authentication/Tokens/TokenManager.swift:82]: I think this should not be discardable right? The whole point of this function to get the token.

File: examples/UberSDK/UberSDKTests/UberCore/KeychainUtilityTests.swift [hunk 5]
```diff
@@ -0,0 +1,77 @@
+//
+//  KeychainUtility.swift
```
Comment (alanpaivaa) [inline examples/UberSDK/UberSDKTests/UberCore/KeychainUtilityTests.swift:2]: NIT: wrong filename

File: examples/UberSDK/UberSDKTests/UberCore/TokenManagerTests.swift [hunk 6]
```diff
@@ -0,0 +1,188 @@
+//
+//  KeychainUtility.swift
```
Comment (alanpaivaa) [inline examples/UberSDK/UberSDKTests/UberCore/TokenManagerTests.swift:2]: wrong file name

File: examples/UberSDK/UberSDKTests/UberCore/KeychainUtilityTests.swift [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+final class KeychainUtilityTests: XCTestCase {
+    
+    func test_save() {
+        let testObject = TestObject(
+            name: "test",
+            value: 5
+        )
+        
+        let keychainUtility = KeychainUtility()
```
Comment (alanpaivaa) [inline examples/UberSDK/UberSDKTests/UberCore/KeychainUtilityTests.swift:37]: IIUC this is accessing the keychain in unit tests. Not sure if we should be doing that. Won't this leave keychain data in the test device which could interfere in the tests?

File: Sources/UberAuth/Button/LoginButton.swift [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+    
+    public weak var dataSource: LoginButtonDataSource?
+    
+    // MARK: Private Properties
+    
+    private var buttonState: State {
+        tokenManager.getToken() != nil ? .loggedIn : .loggedOut
+    }
+    
+    private let tokenManager = TokenManager()
```
[the remark this answers is not in the document]
Comment (mohssenfathi) [reply Sources/UberAuth/Button/LoginButton.swift]: Good call. Fixed.

File: Sources/UberAuth/Button/LoginButton.swift [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+        
+        var title: String {
+            switch self {
+            case .loggedIn:
+                return NSLocalizedString(
+                    "Sign Out",
+                    bundle: .module,
+                    comment: "Login Button Sign Out Description"
+                )
+                .uppercased()
```
[the remark this answers is not in the document]
Comment (mohssenfathi) [reply Sources/UberAuth/Button/LoginButton.swift:151]: Good idea. I think I will leave this one as is until they are used in multiple places.

File: Sources/UberCore/Authentication/Tokens/AccessToken.swift (hunk 1 above)
Comment (mohssenfathi) [reply Sources/UberCore/Authentication/Tokens/AccessToken.swift]: I agree. I ended up keeping the AccessToken class name and changed the property to `tokenString`, as it was before.

File: examples/UberSDK/UberSDKTests/UberCore/KeychainUtilityTests.swift (hunk 7 above)
Comment (mohssenfathi) [reply examples/UberSDK/UberSDKTests/UberCore/KeychainUtilityTests.swift]: Updated the test class to clear the test token on setUp and tearDown

File: Sources/UberAuth/Button/LoginButton.swift [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+        switch buttonState {
+        case .loggedIn:
+            logout()
+        case .loggedOut:
+            login()
+        }
+    }
+    
+    private func login() {
+        let defaultContext = AuthContext(
```
Comment (alanpaivaa) [inline Sources/UberAuth/Button/LoginButton.swift:118]: NIT: could be moved to a constant at class level.

File: Sources/UberAuth/Button/LoginButton.swift [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+            authProvider: .authorizationCode(
+                shouldExchangeAuthCode: true
+            )
+        )
+        let context = dataSource?.authContext(self) ?? defaultContext
+        UberAuth.login(context: context) { [weak self] result in
```
Comment (alanpaivaa) [inline Sources/UberAuth/Button/LoginButton.swift:124]: NIT: weak self likely not needed, since UberAuth is being referred statically

**Review by alanpaivaa — APPROVED** [review]
Thanks for addressing the comments

File: Sources/UberAuth/Button/LoginButton.swift (hunk 11 above)
Comment (mohssenfathi) [reply Sources/UberAuth/Button/LoginButton.swift:124]: Thanks. Will fix in the next PR.

**Pull request merged**